File: core/src/passes/dce.cpp
// core/src/passes/dce.cpp — Dead Code Elimination for the IR.
//
// Two-pass algorithm over a single statement list:
//
//   1. Backward sweep: seed `live` with every Ref appearing as an operand
//      of a side-effecting statement. Then iterate: for each statement
//      whose result is in `live`, also add its operands to `live`. We
//      implement this by walking statements in reverse once, which is
//      sufficient because our IR is in SSA and statements are already in
//      program order (operand defs precede their uses).
//
//   2. Forward filter: emit every statement that is side-effecting, or
//      whose result is in `live`. Drop the rest.
//
// See `prisma/passes.hpp` for the contract on LoadMemTSO handling.

#include "prisma/passes.hpp"

#include <unordered_set>
#include <variant>

namespace prisma::passes {

namespace {

// Classify an Op: true iff removing its statement would NOT change program
// observable behaviour (assuming its result is unused). Side-effecting
// ops stay regardless of whether their result Ref is live.
bool is_pure_for_dce(const ir::Op& op) noexcept {
    return std::visit([](auto const& x) -> bool {
        using T = std::decay_t<decltype(x)>;
        if constexpr (std::is_same_v<T, ir::Constant>) return true;
        else if constexpr (std::is_same_v<T, ir::LoadReg>) return true;
        else if constexpr (std::is_same_v<T, ir::BinOp>) return true;
        else if constexpr (std::is_same_v<T, ir::Compare>) return true;
        else if constexpr (std::is_same_v<T, ir::Select>) return true;
        else if constexpr (std::is_same_v<T, ir::LoadMem>) return true;
        // F2-PS-002: WriteFlags / ReadFlag / WriteFlagsFp /
        // WriteFlagsPtest are pure — they compute a flag-typed Ref
        // from operands without side effects. If no consumer
        // (ReadFlag / CondJumpFlags / lowering-time flag plumbing)
        // reads the result, the write is dead.
        else if constexpr (std::is_same_v<T, ir::WriteFlags>) return true;
        else if constexpr (std::is_same_v<T, ir::ReadFlag>) return true;
        else if constexpr (std::is_same_v<T, ir::WriteFlagsFp>) return true;
        else if constexpr (std::is_same_v<T, ir::WriteFlagsPtest>) return true;
        else if constexpr (std::is_same_v<T, ir::WriteFlagsPtestYmm>) return true;
        else if constexpr (std::is_same_v<T, ir::VecTbl2>) return true;
        else if constexpr (std::is_same_v<T, ir::VecAes>) return true;
        else if constexpr (std::is_same_v<T, ir::JumpReg>) return false;
        // Everything else is impure: StoreReg, StoreMem*, LoadMemTSO,
        // Jump, JumpReg, CondJump, Return, CmpFlags (sets implicit flags),
        // JumpRel / CondJumpRel / CondJumpFlags (control-flow transfer).
        else return false;
    }, op);
}

// Collect every Ref that an Op reads as an operand. The Op's own result
// (if any) is not included — only operands.
void collect_operand_refs(const ir::Op& op, std::unordered_set<ir::Ref>& into) {
    std::visit([&](auto const& x) {
        using T = std::decay_t<decltype(x)>;
        if constexpr (std::is_same_v<T, ir::Constant>) {
            (void)x;
        } else if constexpr (std::is_same_v<T, ir::LoadReg>) {
            (void)x;
        } else if constexpr (std::is_same_v<T, ir::StoreReg>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::BinOp>) {
            into.insert(x.lhs);
            into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::Compare>) {
            into.insert(x.lhs);
            into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::Select>) {
            into.insert(x.true_value);
            into.insert(x.false_value);
        } else if constexpr (std::is_same_v<T, ir::LoadMem>) {
            into.insert(x.addr);
        } else if constexpr (std::is_same_v<T, ir::StoreMem>) {
            into.insert(x.addr);
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::LoadMemTSO>) {
            into.insert(x.addr);
        } else if constexpr (std::is_same_v<T, ir::StoreMemTSO>) {
            into.insert(x.addr);
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::Jump>) {
            (void)x;
        } else if constexpr (std::is_same_v<T, ir::JumpReg>) {
            into.insert(x.target);
        } else if constexpr (std::is_same_v<T, ir::CondJump>) {
            into.insert(x.cond);
        } else if constexpr (std::is_same_v<T, ir::Return>) {
            (void)x;
        } else if constexpr (std::is_same_v<T, ir::CmpFlags>) {
            into.insert(x.lhs);
            into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::JumpRel>) {
            (void)x;
        } else if constexpr (std::is_same_v<T, ir::CondJumpRel>) {
            (void)x;  // no SSA operands; cc + PCs are constants.
        } else if constexpr (std::is_same_v<T, ir::VecBinOp>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecFpBinOp>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecFpScalarBinOp>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::StoreVecReg>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::LoadVec>) {
            into.insert(x.addr);
        } else if constexpr (std::is_same_v<T, ir::StoreVec>) {
            into.insert(x.addr); into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::XmmFromGpr>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::GprFromXmm>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::VecCmp>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecShuffle32x4>) {
            into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::VecUnpack>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecShiftImm>) {
            into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::VecShiftBytes>) {
            into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::IntToFpScalar>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::FpToIntScalar>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::FpCvtScalar>) {
            into.insert(x.lhs); into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::VecShuffle2Src>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecInsertLane>) {
            into.insert(x.lhs_xmm); into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::VecExtractLaneU>) {
            into.insert(x.src_xmm);
        } else if constexpr (std::is_same_v<T, ir::VecMaskMsb>) {
            into.insert(x.src_xmm);
        } else if constexpr (std::is_same_v<T, ir::WriteFlagsFp>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecShuffleH4>) {
            into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::VecMaskFp>) {
            into.insert(x.src_xmm);
        } else if constexpr (std::is_same_v<T, ir::VecFpCompare>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecPshufb>) {
            into.insert(x.src); into.insert(x.mask);
        } else if constexpr (std::is_same_v<T, ir::VecAbs>) {
            into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::VecAlignr>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::VecExtend>) {
            into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::VecFpRound>) {
            into.insert(x.lhs); into.insert(x.src);
        } else if constexpr (std::is_same_v<T, ir::Popcnt>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::Lzcnt>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::Tzcnt>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::VecBlend>) {
            into.insert(x.dst); into.insert(x.src); into.insert(x.mask);
        } else if constexpr (std::is_same_v<T, ir::WriteFlagsPtest>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::WriteFlagsPtestYmm>) {
            into.insert(x.lo_lhs); into.insert(x.lo_rhs);
            into.insert(x.hi_lhs); into.insert(x.hi_rhs);
        } else if constexpr (std::is_same_v<T, ir::VecTbl2>) {
            into.insert(x.src_lo); into.insert(x.src_hi); into.insert(x.idx);
        } else if constexpr (std::is_same_v<T, ir::VecAes>) {
            into.insert(x.src); into.insert(x.key);
        }
        // F2-PS-002: pull flag-related and AVX-256/FMA operand refs
        // through the live-set so DCE can correctly see what's used.
        else if constexpr (std::is_same_v<T, ir::WriteFlags>) {
            into.insert(x.lhs); into.insert(x.rhs);
        } else if constexpr (std::is_same_v<T, ir::ReadFlag>) {
            into.insert(x.flags);
        } else if constexpr (std::is_same_v<T, ir::CondJumpFlags>) {
            into.insert(x.flags);
        } else if constexpr (std::is_same_v<T, ir::StoreVecRegHi>) {
            into.insert(x.value);
        } else if constexpr (std::is_same_v<T, ir::VecFpFma>) {
            into.insert(x.a); into.insert(x.b); into.insert(x.c);
        } else if constexpr (std::is_same_v<T, ir::VecFpScalarFma>) {
            into.insert(x.a); into.insert(x.b); into.insert(x.c);
            into.insert(x.scalar_upper);
        }
        // LoadVecRegHi and the various producer-only Load* ops have no
        // operand refs and need no entry here.
    }, op);
}

}  // namespace

std::vector<ir::Stmt>
dead_code_eliminate(const std::vector<ir::Stmt>& stmts) {
    // -------- Pass 1: collect live refs by walking statements in reverse.
    std::unordered_set<ir::Ref> live;

    for (auto it = stmts.rbegin(); it != stmts.rend(); ++it) {
        const ir::Stmt& s = *it;
        const bool has_result = s.result.has_value();
        const bool impure     = !is_pure_for_dce(s.op);
        const bool result_is_live = has_result && live.count(*s.result) != 0;

        // A statement's operands become live if:
        //   * the op is side-effecting (we must keep it, so its operands
        //     are needed), OR
        //   * its result is already live (so computing it requires the
        //     operands).
        if (impure || result_is_live) {
            collect_operand_refs(s.op, live);
        }
    }

    // -------- Pass 2: filter.
    std::vector<ir::Stmt> out;
    out.reserve(stmts.size());
    for (const ir::Stmt& s : stmts) {
        const bool impure = !is_pure_for_dce(s.op);
        const bool result_is_live =
            s.result.has_value() && live.count(*s.result) != 0;

        if (impure || result_is_live) {
            out.push_back(s);
        }
    }
    return out;
}

}  // namespace prisma::passes

File: shell/prisma-passes/src/dce.rs
//! Dead code elimination.
//!
//! Mirrors C++ `dead_code_eliminate`: a reverse liveness sweep seeds the live
//! set from the operands of side-effecting statements and propagates through
//! statements whose result is live; a forward filter then drops pure
//! statements whose result is unused. SSA + program order make a single
//! reverse pass sufficient (operand defs precede uses).
//!
//! The purity classification mirrors the C++ allowlist exactly so the Rust and
//! C++ pipelines stay differentially equivalent.

use std::collections::HashSet;

use prisma_ir::{BasicBlock, Function, Op, Ref, Stmt};

use crate::Pass;

/// True iff dropping the statement (when its result is unused) cannot change
/// observable behaviour. Side-effecting ops stay regardless of liveness.
fn is_pure_for_dce(op: &Op) -> bool {
    matches!(
        op,
        Op::Constant(_)
            | Op::LoadReg(_)
            | Op::BinOp(_)
            | Op::Compare(_)
            | Op::Select(_)
            | Op::LoadMem(_)
            | Op::WriteFlags(_)
            | Op::ReadFlag(_)
            | Op::WriteFlagsFp(_)
            | Op::VecConstant(_)
            | Op::LoadVecReg(_)
            | Op::LoadVecRegHi(_)
            | Op::WriteFlagsPtest(_)
            | Op::WriteFlagsPtestYmm(_)
            | Op::VecTbl2(_)
            | Op::VecAes(_)
            | Op::VecAesKeygenAssist(_)
            | Op::VecSha(_)
            | Op::Bswap(_)
            | Op::Crc32c(_)
            | Op::VecGather(_)
            | Op::X87Load(_)
            | Op::LoadCarry(_)
            | Op::ReadCarryOut(_)
    )
}

/// Insert every SSA `Ref` the op reads as an operand into `into`.
///
/// Arms are kept one-per-variant (mirroring the C++ visitor) even when their
/// bodies coincide, so the operand map stays auditable against `ir.hpp`.
#[allow(clippy::too_many_lines, clippy::match_same_arms)]
fn collect_operand_refs(op: &Op, into: &mut HashSet<Ref>) {
    match op {
        // No operand refs.
        Op::Constant(_)
        | Op::LoadReg(_)
        | Op::LoadSegBase(_)
        | Op::Jump(_)
        | Op::Return(_)
        | Op::JumpRel(_)
        | Op::CondJumpRel(_)
        | Op::CallRel(_)
        | Op::RetAdjusted(_)
        | Op::Cpuid(_)
        | Op::Xgetbv(_)
        | Op::Rdtsc(_)
        | Op::Syscall(_)
        | Op::Trap(_)
        | Op::Fence(_)
        | Op::GuestPc(_)
        | Op::InlineAsm(_)
        | Op::FpConstant(_)
        | Op::RspAdjust(_)
        | Op::VecConstant(_)
        | Op::LoadVecReg(_)
        | Op::LoadVecRegHi(_)
        | Op::X87Load(_)
        | Op::X87Pop(_)
        | Op::LoadCarry(_) => {}

        Op::ReadCarryOut(x) => {
            into.insert(x.flags);
        }
        Op::StoreCarry(x) => {
            into.insert(x.value);
        }
        Op::StoreReg(x) => {
            into.insert(x.value);
        }
        Op::BinOp(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::Compare(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::Select(x) => {
            into.insert(x.true_value);
            into.insert(x.false_value);
        }
        Op::LoadMem(x) => {
            into.insert(x.addr);
        }
        Op::StoreMem(x) => {
            into.insert(x.addr);
            into.insert(x.value);
        }
        Op::LoadMemTSO(x) => {
            into.insert(x.addr);
        }
        Op::StoreMemTSO(x) => {
            into.insert(x.addr);
            into.insert(x.value);
        }
        Op::CondJump(x) => {
            into.insert(x.cond);
        }
        Op::CmpFlags(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::AluFlags(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::JumpReg(x) => {
            into.insert(x.target);
        }
        Op::CallReg(x) => {
            into.insert(x.target);
        }
        Op::Extend(x) => {
            into.insert(x.value);
        }
        Op::Truncate(x) => {
            into.insert(x.value);
        }
        Op::FpBinOp(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::WriteFlags(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::ReadFlag(x) => {
            into.insert(x.flags);
        }
        Op::CondJumpFlags(x) => {
            into.insert(x.flags);
        }
        Op::VecBinOp(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::StoreVecReg(x) => {
            into.insert(x.value);
        }
        Op::LoadVec(x) => {
            into.insert(x.addr);
        }
        Op::StoreVec(x) => {
            into.insert(x.addr);
            into.insert(x.value);
        }
        Op::VecFpBinOp(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::VecFpScalarBinOp(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::XmmFromGpr(x) => {
            into.insert(x.value);
        }
        Op::GprFromXmm(x) => {
            into.insert(x.value);
        }
        Op::VecCmp(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::VecShuffle32x4(x) => {
            into.insert(x.src);
        }
        Op::VecUnpack(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::VecShiftImm(x) => {
            into.insert(x.src);
        }
        Op::VecShiftBytes(x) => {
            into.insert(x.src);
        }
        Op::IntToFpScalar(x) => {
            into.insert(x.value);
        }
        Op::FpToIntScalar(x) => {
            into.insert(x.value);
        }
        Op::FpCvtScalar(x) => {
            into.insert(x.lhs);
            into.insert(x.src);
        }
        Op::VecShuffle2Src(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::VecInsertLane(x) => {
            into.insert(x.lhs_xmm);
            into.insert(x.value);
        }
        Op::VecExtractLaneU(x) => {
            into.insert(x.src_xmm);
        }
        Op::VecMaskMsb(x) => {
            into.insert(x.src_xmm);
        }
        Op::WriteFlagsFp(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::VecShuffleH4(x) => {
            into.insert(x.src);
        }
        Op::VecMaskFp(x) => {
            into.insert(x.src_xmm);
        }
        Op::VecFpCompare(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::VecPshufb(x) => {
            into.insert(x.src);
            into.insert(x.mask);
        }
        Op::VecAbs(x) => {
            into.insert(x.src);
        }
        Op::VecAlignr(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::VecExtend(x) => {
            into.insert(x.src);
        }
        Op::VecFpRound(x) => {
            into.insert(x.lhs);
            into.insert(x.src);
        }
        Op::Popcnt(x) => {
            into.insert(x.value);
        }
        Op::Lzcnt(x) => {
            into.insert(x.value);
        }
        Op::Tzcnt(x) => {
            into.insert(x.value);
        }
        Op::WriteFlagsPopcnt(x) => {
            into.insert(x.src);
        }
        Op::WriteFlagsCountZero(x) => {
            into.insert(x.src);
            into.insert(x.result);
        }
        Op::VecBlend(x) => {
            into.insert(x.dst);
            into.insert(x.src);
            into.insert(x.mask);
        }
        Op::WriteFlagsPtest(x) => {
            into.insert(x.lhs);
            into.insert(x.rhs);
        }
        Op::WriteFlagsPtestYmm(x) => {
            into.insert(x.lo_lhs);
            into.insert(x.lo_rhs);
            into.insert(x.hi_lhs);
            into.insert(x.hi_rhs);
        }
        Op::VecTbl2(x) => {
            into.insert(x.src_lo);
            into.insert(x.src_hi);
            into.insert(x.idx);
        }
        Op::VecAes(x) => {
            into.insert(x.src);
            into.insert(x.key);
        }
        Op::VecAesKeygenAssist(x) => {
            into.insert(x.src);
        }
        Op::VecSha(x) => {
            into.insert(x.a);
            into.insert(x.b);
            into.insert(x.wk);
        }
        Op::Bswap(x) => {
            into.insert(x.value);
        }
        Op::Crc32c(x) => {
            into.insert(x.crc);
            into.insert(x.data);
        }
        Op::VecGather(x) => {
            into.insert(x.base);
            into.insert(x.index);
            into.insert(x.mask);
            into.insert(x.prev);
        }
        Op::StoreVecRegHi(x) => {
            into.insert(x.value);
        }
        Op::VecFpFma(x) => {
            into.insert(x.a);
            into.insert(x.b);
            into.insert(x.c);
        }
        Op::VecFpScalarFma(x) => {
            into.insert(x.a);
            into.insert(x.b);
            into.insert(x.c);
            into.insert(x.scalar_upper);
        }
        Op::RepStos(_) | Op::RepMovs(_) => {}
        Op::X87Store(x) => {
            into.insert(x.value);
        }
        Op::X87Push(x) => {
            into.insert(x.value);
        }
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct Dce;

impl Dce {
    #[must_use]
    pub const fn new() -> Self {
        Self
    }
}

impl Pass for Dce {
    fn name(&self) -> &'static str {
        "dead_code_eliminate"
    }

    fn run(&self, func: Function) -> Function {
        dead_code_eliminate(func)
    }
}

/// Dead code elimination with **function-global** liveness.
///
/// SSA refs are function-scoped, so a value defined in one block may be used
/// in a later block. The C++ pass runs over a single flat statement vector; a
/// per-block reverse sweep would miss cross-block uses and wrongly drop a live
/// def. We therefore sweep liveness across all blocks (in program order,
/// reversed) into one shared live set before filtering. Within the function,
/// SSA + program order guarantees defs precede uses, so a single reverse pass
/// over the concatenation suffices — exactly mirroring the C++ flat scan.
pub fn dead_code_eliminate(func: Function) -> Function {
    // Pass 1: reverse liveness sweep over every statement in the function.
    let mut live: HashSet<Ref> = HashSet::new();
    for block in func.blocks.iter().rev() {
        for stmt in block.stmts.iter().rev() {
            let impure = !is_pure_for_dce(&stmt.op);
            let result_is_live = stmt.result.is_some_and(|r| live.contains(&r));
            if impure || result_is_live {
                collect_operand_refs(&stmt.op, &mut live);
            }
        }
    }

    // Pass 2: forward filter per block against the global live set.
    let blocks = func
        .blocks
        .into_iter()
        .map(|block| {
            let stmts = block
                .stmts
                .into_iter()
                .filter(|stmt| {
                    let impure = !is_pure_for_dce(&stmt.op);
                    let result_is_live = stmt.result.is_some_and(|r| live.contains(&r));
                    impure || result_is_live
                })
                .collect();
            BasicBlock {
                id: block.id,
                stmts,
            }
        })
        .collect();

    Function {
        blocks,
        entry: func.entry,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use prisma_ir::{BinOp, BinOpKind, Constant, Gpr, OpSize, StoreReg};

    #[test]
    fn dead_constant_is_dropped() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(
                        Some(0),
                        Op::Constant(Constant {
                            value: 7,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(None, Op::Return(prisma_ir::Return)),
                ],
            }],
        };
        let out = dead_code_eliminate(func);
        // The unused constant is removed; only Return remains.
        assert_eq!(out.blocks[0].stmts.len(), 1);
        assert!(matches!(out.blocks[0].stmts[0].op, Op::Return(_)));
    }

    #[test]
    fn live_chain_is_kept() {
        // r0 = const; r1 = r0+r0; StoreReg rax, r1  (r1, r0 live)
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(
                        Some(0),
                        Op::Constant(Constant {
                            value: 7,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        Some(1),
                        Op::BinOp(BinOp {
                            op: BinOpKind::Add,
                            lhs: 0,
                            rhs: 0,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        None,
                        Op::StoreReg(StoreReg {
                            reg: Gpr::Rax,
                            value: 1,
                            size: OpSize::I64,
                        }),
                    ),
                ],
            }],
        };
        let out = dead_code_eliminate(func.clone());
        assert_eq!(out, func);
    }

    #[test]
    fn partially_dead_is_pruned() {
        // r0 used by store; r1 dead.
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(
                        Some(0),
                        Op::Constant(Constant {
                            value: 1,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        Some(1),
                        Op::Constant(Constant {
                            value: 2,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        None,
                        Op::StoreReg(StoreReg {
                            reg: Gpr::Rax,
                            value: 0,
                            size: OpSize::I64,
                        }),
                    ),
                ],
            }],
        };
        let out = dead_code_eliminate(func);
        assert_eq!(out.blocks[0].stmts.len(), 2);
        // r1 constant dropped, r0 + store kept.
        assert!(matches!(out.blocks[0].stmts[1].op, Op::StoreReg(_)));
    }

    #[test]
    fn idempotent() {
        let func = Function {
            entry: 0,
            blocks: vec![BasicBlock {
                id: 0,
                stmts: vec![
                    Stmt::new(
                        Some(0),
                        Op::Constant(Constant {
                            value: 1,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        Some(1),
                        Op::Constant(Constant {
                            value: 2,
                            size: OpSize::I64,
                        }),
                    ),
                    Stmt::new(
                        None,
                        Op::StoreReg(StoreReg {
                            reg: Gpr::Rax,
                            value: 0,
                            size: OpSize::I64,
                        }),
                    ),
                ],
            }],
        };
        let once = dead_code_eliminate(func);
        let twice = dead_code_eliminate(once.clone());
        assert_eq!(once, twice);
    }

    #[test]
    fn cross_block_use_keeps_def() {
        // r0 defined (pure) in block 0, used by a StoreReg in block 1.
        // A per-block liveness sweep would wrongly drop r0; the global
        // sweep must keep it (regression for the Codex BLOCKER).
        let func = Function {
            entry: 0,
            blocks: vec![
                BasicBlock {
                    id: 0,
                    stmts: vec![Stmt::new(
                        Some(0),
                        Op::Constant(Constant {
                            value: 7,
                            size: OpSize::I64,
                        }),
                    )],
                },
                BasicBlock {
                    id: 1,
                    stmts: vec![Stmt::new(
                        None,
                        Op::StoreReg(StoreReg {
                            reg: Gpr::Rax,
                            value: 0,
                            size: OpSize::I64,
                        }),
                    )],
                },
            ],
        };
        let out = dead_code_eliminate(func.clone());
        assert_eq!(out, func, "cross-block live def must not be dropped");
    }
}
